### PR TITLE
Skip directory symlink recursion on TarFile archive creation

### LINKED ISSUE #74214: Tar: behavior for symbolic links to directories
Unlike `tar`, currently these links are being followed.
If the file linked to is in the archive it will be duplicated. If it is outside the archived directory it will be included as if it were.
Symbolic links to directories can also cause loops which will cause the archive creation to hang.

Can we behave like `tar` and not follow symlinks by default?
`tar` does provide an option to follow the links (`--dereference`).

cc @carlossanlop @dotnet/area-system-io

### PR DESCRIPTION
**Pull request by carlossanlop**
Fixes https://github.com/dotnet/runtime/issues/74214

### CODE REVIEW

File: src/libraries/System.Formats.Tar/tests/TarFile/TarFile.CreateFromDirectory.File.Tests.cs [hunk 1]
```diff
@@ -193,5 +193,65 @@ public void IncludeAllSegmentsOfPath(bool includeBaseDirectory)
...
+
+            TarFile.CreateFromDirectory(sourceDirectoryName, destinationArchive, includeBaseDirectory: false);
+
+            using FileStream archiveStream = File.OpenRead(destinationArchive);
+            using (TarReader reader = new(archiveStream, leaveOpen: false))
+            {
+                TarEntry entry = reader.GetNextEntry();
+                Assert.NotNull(entry);
+                Assert.Equal("subDirectory/", entry.Name);
+
```
Comment (Jozkee) [inline src/libraries/System.Formats.Tar/tests/TarFile/TarFile.CreateFromDirectory.File.Tests.cs:223]:
```suggestion
                Assert.Equal("subDirectory/", entry.Name);
                Assert.Equal(TarEntyrType.SymbolicLink, entry.EntryType); 
```

File: src/libraries/System.Formats.Tar/tests/TarFile/TarFile.CreateFromDirectory.File.Tests.cs [hunk 2]
```diff
@@ -193,5 +193,65 @@ public void IncludeAllSegmentsOfPath(bool includeBaseDirectory)
...
+            string externalDirectory = Path.Join(root.Path, "externalDirectory");
+            Directory.CreateDirectory(externalDirectory);
+
+            string subDirectory = Path.Join(externalDirectory, "subDirectory");
+            Directory.CreateDirectory(subDirectory);
+
+            string sourceDirectoryName = Path.Join(root.Path, "baseDirectory");
+            Directory.CreateSymbolicLink(sourceDirectoryName, externalDirectory);
+
+            TarFile.CreateFromDirectory(sourceDirectoryName, destinationArchive, includeBaseDirectory: true); // Base directory is a symlink, do not recurse
```
Comment (Jozkee) [inline src/libraries/System.Formats.Tar/tests/TarFile/TarFile.CreateFromDirectory.File.Tests.cs:244]: Is this how `tar` works? Doesn't it resolve the link when it is the base directory?

Comment (carlossanlop) [reply src/libraries/System.Formats.Tar/tests/TarFile/TarFile.CreateFromDirectory.File.Tests.cs:244]: Yes:

```bash
$ cd ~
$ mkdir root
$ cd root
$ mkdir externalDirectory
$ mkdir externalDirectory/baseDirectory
$ ln -s externalDirectory/ baseDirectory
$ ls
baseDirectory externalDirectory
ls externalDirectory
subDirectory
$ ls baseDirectory
subDirectory
$ tar -cvf destination.tar baseDirectory
baseDirectory
```

If I open `destination.tar` with an hex editor, you'll see a single entry, the base directory, without following the symlink:

image

File: src/libraries/System.Formats.Tar/tests/TarFile/TarFile.CreateFromDirectory.File.Tests.cs (hunk 1 above)
Comment (carlossanlop) [reply src/libraries/System.Formats.Tar/tests/TarFile/TarFile.CreateFromDirectory.File.Tests.cs:223]: Working on this for all the 4 tests.

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarFile.cs [hunk 3]
```diff
@@ -325,18 +334,44 @@ await using (writer.ConfigureAwait(false))
...
+        {
+            return new FileSystemEnumerable<FileSystemInfo>(
+                directory: sourceDirectoryName,
+                transform: (ref FileSystemEntry entry) => entry.ToFileSystemInfo(),
+                options: new EnumerationOptions()
+                {
+                    RecurseSubdirectories = true
+                })
+            {
+                ShouldRecursePredicate = IsDirectorySymlink
```
Comment (danmoseley) [inline src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarFile.cs:369]: This reads like it only recurses symlinks not the reverse?

File: src/libraries/System.Formats.Tar/tests/TarFile/TarFile.CreateFromDirectoryAsync.File.Tests.cs [hunk 4]
```diff
@@ -237,5 +237,67 @@ await using (TarReader reader = new TarReader(fileStream))
...
+            string sourceDirectoryName = Path.Join(root.Path, "baseDirectory");
+            Directory.CreateDirectory(sourceDirectoryName);
+
+            string subDirectory = Path.Join(sourceDirectoryName, "subDirectory");
+            Directory.CreateSymbolicLink(subDirectory, externalDirectory); // Should not recurse here
+
+            await TarFile.CreateFromDirectoryAsync(sourceDirectoryName, destinationArchive, includeBaseDirectory: false);
+
+            await using FileStream archiveStream = File.OpenRead(destinationArchive);
+            await using (TarReader reader = new(archiveStream, leaveOpen: false))
```
Comment (danmoseley) [inline src/libraries/System.Formats.Tar/tests/TarFile/TarFile.CreateFromDirectoryAsync.File.Tests.cs:262]: Nit, if you're using single line using above, you can use it for TarReader

**Review by danmoseley — COMMENTED** [review]
With this change can we read anything new in the new corpus ?

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarFile.cs (hunk 3 above)
Comment (carlossanlop) [reply src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarFile.cs]: Good point. Fixed.

File: src/libraries/System.Formats.Tar/tests/TarFile/TarFile.CreateFromDirectoryAsync.File.Tests.cs (hunk 4 above)
Comment (carlossanlop) [reply src/libraries/System.Formats.Tar/tests/TarFile/TarFile.CreateFromDirectoryAsync.File.Tests.cs]: Fixed.

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarFile.cs (hunk 3 above)
Comment (danmoseley) [reply src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarFile.cs]: Did a test fail because of that?

**Comment by carlossanlop** [comment]
> With this change can we read anything new in the new corpus ? 

I looked through the new assets, didn't find any that could repro this issue. This was originally reported by @tmds.

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarFile.cs [hunk 5]
```diff
@@ -278,12 +279,20 @@ private static void CreateFromDirectoryInternal(string sourceDirectoryName, Stre
                 DirectoryInfo di = new(sourceDirectoryName);
                 string basePath = GetBasePathForCreateFromDirectory(di, includeBaseDirectory);
 
+                bool skipBaseDirRecursion = false;
                 if (includeBaseDirectory)
                 {
                     writer.WriteEntry(di.FullName, GetEntryNameForBaseDirectory(di.Name));
+                    skipBaseDirRecursion = (di.Attributes & FileAttributes.ReparsePoint) != 0;
```
Comment (Jozkee) [inline src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarFile.cs:286]: you don't need `skipBaseDirRecursion`.
```suggestion
                    if ((di.Attributes & FileAttributes.ReparsePoint) != 0)
                    {
                        return;
                    }
```

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarFile.cs [hunk 6]
```diff
@@ -325,18 +334,44 @@ await using (writer.ConfigureAwait(false))
                 DirectoryInfo di = new(sourceDirectoryName);
                 string basePath = GetBasePathForCreateFromDirectory(di, includeBaseDirectory);
 
+                bool skipBaseDirRecursion = false;
                 if (includeBaseDirectory)
                 {
                     await writer.WriteEntryAsync(di.FullName, GetEntryNameForBaseDirectory(di.Name), cancellationToken).ConfigureAwait(false);
+                    skipBaseDirRecursion = (di.Attributes & FileAttributes.ReparsePoint) != 0;
```
Comment (Jozkee) [inline src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarFile.cs:341]: ditto

**Review by jozkee — APPROVED** [review]
Otherwise; LGTM.

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarFile.cs (hunk 5 above)
Comment (carlossanlop) [reply src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarFile.cs:286]: Both suggestions can be fixed in my next PR.

**Comment by carlossanlop** [comment]
Test failures unrelated:
System.Net.Quic.Functional.Tests - https://github.com/dotnet/runtime/issues/73688
System.Net.Http.Functional.Tests - https://github.com/dotnet/runtime/issues/74162

**Comment by carlossanlop** [comment]
@lewing I am including this backport in this draft: https://github.com/dotnet/runtime/pull/74414

**Pull request merged**